### PR TITLE
feat(tracking): add telemetry for frontend events

### PR DESCRIPTION
**Pull request by aditya-radhakrishnan**
## Checklist
- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [x] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [x] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: datahub-frontend/app/client/AuthServiceClient.java [hunk 1]
```diff
@@ -259,6 +260,44 @@ public boolean verifyNativeUserCredentials(@Nonnull final String userUrn, @Nonnu
...
+      CloseableHttpResponse response = httpClient.execute(request);
+      final HttpEntity entity = response.getEntity();
+
+      if (response.getStatusLine().getStatusCode() != HttpStatus.SC_OK || entity == null) {
+        throw new RuntimeException(
+            String.format("Bad response from the Metadata Service: %s %s", response.getStatusLine().toString(),
+                response.getEntity().toString()));
+      }
+    } catch (Exception e) {
+      throw new RuntimeException("Failed to verify credentials for user", e);
```
Comment (gabe-lyons) [inline datahub-frontend/app/client/AuthServiceClient.java:291]: `Failed to verify credentials for user` is this the right error message?

File: datahub-frontend/app/client/AuthServiceClient.java [hunk 2]
```diff
@@ -259,6 +260,44 @@ public boolean verifyNativeUserCredentials(@Nonnull final String userUrn, @Nonnu
...
+    } finally {
+      try {
+        httpClient.close();
```
Comment (gabe-lyons) [inline datahub-frontend/app/client/AuthServiceClient.java:294]: do we need to open & close this client each time we make a request? theres no re-usable client?

File: metadata-service/auth-impl/src/main/java/com/datahub/telemetry/TrackingService.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    // Otherwise, return the existing client id from the metadata store
+    RecordTemplate clientIdTemplate = _entityService.getLatestAspect(clientIdUrn, CLIENT_ID_ASPECT);
+    // Should always be present here from above, so no need for null check
+    _clientId = ((TelemetryClientId) clientIdTemplate).getClientId();
+    return _clientId;
+  }
+
+  @Nonnull
+  JSONObject sanitizeEvent(@Nonnull final JsonNode event) {
+    final JSONObject eventObj;
```
Comment (gabe-lyons) [inline metadata-service/auth-impl/src/main/java/com/datahub/telemetry/TrackingService.java:117]: for the sake of no confusion, would be good to call this `unsanitizedEventObj`

Comment (aditya-radhakrishnan) [reply metadata-service/auth-impl/src/main/java/com/datahub/telemetry/TrackingService.java:117]: Good call, renamed

File: datahub-frontend/app/client/AuthServiceClient.java (hunk 2 above)
Comment (aditya-radhakrishnan) [reply datahub-frontend/app/client/AuthServiceClient.java:294]: Refactored to use one client

File: metadata-service/auth-servlet-impl/src/main/java/com/datahub/authentication/AuthServiceController.java [hunk 4]
```diff
@@ -307,6 +313,36 @@ CompletableFuture<ResponseEntity<String>> verifyNativeUserCredentials(final Http
...
+   */
+  @PostMapping(value = "/track", produces = "application/json;charset=utf-8")
+  CompletableFuture<ResponseEntity<String>> track(final HttpEntity<String> httpEntity) {
+    String jsonStr = httpEntity.getBody();
+    ObjectMapper mapper = new ObjectMapper();
+    JsonNode bodyJson;
+    try {
+      bodyJson = mapper.readTree(jsonStr);
+    } catch (JsonProcessingException e) {
+      log.error(String.format("Failed to parse json while attempting to create native user %s", jsonStr));
```
Comment (gabe-lyons) [inline metadata-service/auth-servlet-impl/src/main/java/com/datahub/authentication/AuthServiceController.java:327]: is this comment right?

**Review by gabe-lyons — APPROVED** [review]

**Pull request merged**